src/rmp/src/genetic_strategy.cpp

```diff
@@ -0,0 +1,171 @@
+    const std::vector<GiaOp>& all_ops,
+    const std::vector<sta::Vertex*>& candidate_vertices,
+    cut::AbcLibrary& abc_library,
+    sta::dbSta* sta,
+    utl::UniqueName& name_generator,
+    rsz::Resizer* resizer,
+    utl::Logger* logger)
+{
+  // Initial solution and slack
+  debugPrint(logger,
+             RMP,
+             "population",
+             1,
+             "Generating and evaluating the initial population");
+  std::vector<SolutionSlack> population(pop_size_);
+  for (auto& ind : population) {
+    ind.solution_.reserve(initial_ops_);
+    for (size_t i = 0; i < initial_ops_; i++) {
+      ind.solution_.push_back(all_ops[random_() % all_ops.size()]);
+    }
+  }
+
+  for (auto& candidate : population) {
+    candidate.Evaluate(
+        candidate_vertices, abc_library, corner_, sta, name_generator, logger);
+
+    debugPrint(logger, RMP, "genetic", 1, candidate.toString());
+  }
+
+  for (unsigned i = 0; i < iterations_; i++) {
+    logger->info(RMP, 65, "Resynthesis: Iteration {} of genetic algorithm", i);
+    unsigned generation_size = population.size();
+    // Crossover
+    unsigned cross_size = std::max<unsigned>(cross_prob_ * generation_size, 1);
+    for (unsigned j = 0; j < cross_size; j++) {
+      auto rand1 = random_() % generation_size;
+      auto rand2 = random_() % generation_size;
+      if (rand1 == rand2) {
+        continue;
+      }
+      SolutionSlack::Type& parent1_sol = population[rand1].solution_;
+      SolutionSlack::Type& parent2_sol = population[rand2].solution_;
+      SolutionSlack::Type child_sol(
+          parent1_sol.begin(), parent1_sol.begin() + parent1_sol.size() / 2);
+      child_sol.insert(child_sol.end(),
+                       parent2_sol.begin() + parent2_sol.size() / 2,
+                       parent2_sol.end());
+      SolutionSlack child_sol_slack;
+      child_sol_slack.solution_ = std::move(child_sol);
+      population.emplace_back(child_sol_slack);
+    }
+    // Mutations
+    unsigned mut_size = std::max<unsigned>(mut_prob_ * generation_size, 1);
+    for (unsigned j = 0; j < mut_size; j++) {
+      SolutionSlack sol_slack;
+      auto rand = random_() % generation_size;
+      sol_slack.solution_
+          = population[rand].RandomNeighbor(all_ops, logger, random_);
+      population.emplace_back(sol_slack);
+    }
+    removeDuplicates(population, logger);
+    // Evaluation
+    for (auto& sol_slack : population) {
+      if (sol_slack.worst_slack_) {
+        continue;
+      }
+      sol_slack.Evaluate(candidate_vertices,
+                         abc_library,
+                         corner_,
+                         sta,
+                         name_generator,
+                         logger);
+    }
+    // Selection
+    std::ranges::sort(population, std::less{});
+    std::vector<SolutionSlack> newPopulation;
+    newPopulation.reserve(pop_size_);
+    for (int j = 0; j < pop_size_; j++) {
+      std::vector<size_t> tournament(tourn_size_);
```

Comment:
If I misunderstood above then we can at least do something like this. Extract the tournament vector from the `j` loop.   Make another vector called indices that has indices 0..pop_size_-1.   Then you can do:
```
tournament_candidates.clear();
std::sample(indices.begin(), indices.end(),
                    std::back_inserter(tournament_candidates),
                    tourn_size_,
                    random_);
```
The indices will come out already sorted so you won't need to sort and erase.   The only difference is that tournament will always have size tourn_size_ since std::sample doesn't produce duplicates.